### PR TITLE
PIC-380 add breaches to convictions in probation record response

### PR DESCRIPTION
**Pull request by tomsmyers**

### CODE REVIEW

File: src/main/java/uk/gov/justice/probation/courtcaseservice/restclient/communityapi/mapper/interventions/BreachMapper.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.stream.Collectors;
+
+import org.springframework.stereotype.Component;
+import uk.gov.justice.probation.courtcaseservice.service.model.Breach;
+import uk.gov.justice.probation.courtcaseservice.restclient.communityapi.model.CommunityApiNsi;
+import uk.gov.justice.probation.courtcaseservice.restclient.communityapi.model.CommunityApiNsiResponse;
+
+@Component
+public class BreachMapper {
+    public Breach breachFrom(CommunityApiNsi nsi) {
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/restclient/communityapi/mapper/interventions/BreachMapper.java:13]: Would expect a unit test class direct on this. It's probably effectively tested by integration tests higher up additionally but please add

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/Breach.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import java.time.LocalDate;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+@ApiModel("Breach")
+@Getter
+@Builder
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class Breach {
+    private String id;
```
Comment (chris-faulkner-mojdt) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/Breach.java:16]: Why string here ? The community API has the nsi id as a long ? Also, lower down, the treatment of convictionId as a string is an overhead in the comparable implementation (although not part of this change)

File: src/test/java/uk/gov/justice/probation/courtcaseservice/service/OffenderServiceTest.java [hunk 3]
```diff
@@ -83,6 +92,7 @@ void beforeEach() {
     void whenGetOffender_returnOffenderWithConvictionsDocumentsNotFiltered() {
         when(offenderRestClient.getProbationRecordByCrn(CRN)).thenReturn(Mono.just(ProbationRecord.builder().crn(CRN).build()));
         when(offenderRestClient.getConvictionsByCrn(CRN)).thenReturn(Mono.just(singletonList(conviction)));
+        when(offenderRestClient.getBreaches(CRN, CONVICTION_ID)).thenReturn(Mono.just(singletonList(breach)));
```
Comment (chris-faulkner-mojdt) [inline src/test/java/uk/gov/justice/probation/courtcaseservice/service/OffenderServiceTest.java:95]: Would like to see the Mockito verify call to prove that this happened. Mockito runs in non-strict mode and it's good to know that each and every "when" is necessary and doing as expected.

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/Breach.java (hunk 2 above)
Comment (tomsmyers) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/Breach.java:16]: i chose String because that seems to be the representation for ids in other models e.g. `documentId` `convictionId`.

agree there is some overhead in converting convictionId for comparisons, but i didn't want to change the existing conviction model.

File: src/test/java/uk/gov/justice/probation/courtcaseservice/service/OffenderServiceTest.java (hunk 3 above)
Comment (tomsmyers) [reply src/test/java/uk/gov/justice/probation/courtcaseservice/service/OffenderServiceTest.java:95]: can do, although mockito does enforce this anyway. if you add a stub for something that isn't called the test fails with an exception like this:

```
org.mockito.exceptions.misusing.UnnecessaryStubbingException: 
Unnecessary stubbings detected.
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at uk.gov.justice.probation.courtcaseservice.service.OffenderServiceTest.givenOffenderNotFound_whenGetOffender_thenThrowException(OffenderServiceTest.java:144)
```

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/Breach.java (hunk 2 above)
Comment (chris-faulkner-mojdt) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/Breach.java:16]: OK - understand the reason not to change the conviction. documentId really is a string - it's a GUID type thing. I'd like to see us using a long unless there's a reason for string

Comment (tomsmyers) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/model/Breach.java:16]: done
